### ISSUE TITLE
Align the decimal place and using 3 signification figures

### ISSUE THREAD

**Issue by shajoezhu** [insightsengineering/formatters#26]
Hi @gmbecker 

Our team currently is dealing with request, I was wondering could you help please, see example below.

![image](https://user-images.githubusercontent.com/3692541/167969307-3347566a-4e92-41a3-a7a9-26cd279ba93e.png)

The 3 signification figures request is now two parts

1. Align the decimal places in the column.
2. Then take the significant digits.

Could you take a look at this please. Thanks so much!

Original request from https://github.com/insightsengineering/tern/issues/544

**Comment by gmbecker** [comment]
So there are a number of things going on here. 

first is decimal alignment. This is on our roadmap but its a heavier lift than it seems because we don't have a way to specify it AND the number of decimal places, even once the logic is in, which is also non-trivial. 

The second the decimal place even being there when there is nothing after it. That...doesn't strike me as a good behavior, to be honest, so I need to hear more about why you would want that.

The third is determining singificant digits empirically, which we have some thoughts about but its another non-trivial thing because it requires the full vector of raw data so it can't actually happen at formatting time.

**Comment by gmbecker** [comment]
@waddella for discussion re place in the roadmap

**Comment by shajoezhu** [comment]
There is some limitations on `signif` function, `x` needs to be a single number, otherwise dont work

```
> x = c(0.0011111, 0.111111)
> signif(x, digits = 3)
[1] 0.00111 0.11100
```

**Comment by qiliu1013** [comment]
@legrasv and @shajoezhu, we don't apply 3 sig. figures in all the numbers. Some need to be one decimal place, and some need 3 sig. figure depending on which statistics or parameters we are looking at. For PKCT01, It was mandatory that the mean, SD, SE, Geo. mean are 3 sig. figures, and CV%, CV% Geo. mean are one one decimal place. It would be great if we can have flexibility to decide whether it should be one decimal place or 3 sig. figures.

**Comment by legrasv** [comment]
Yes sure, if we can have a function and choose the number of digits, that would be great

**Comment by gmbecker** [comment]
@qiliu1013 That won't be an issue, formats are already not required to be uniform across entire rows (or entire columns, though here you do happen to want that).

You can also do significant figures via custom formatting function now, as a workaround:

```
threesigfmt <- function(x,...) {
    as.character(signif(x, 3))
}

fun_a <- function(x, ...) in_rows(readout = mean(x), .formats = list(readout = threesigfmt))

fun_b <- function(x, ...) in_rows(readout = range(x), .formats = list(readout = "xx.x - xx.x"))

lyt3 <- basic_table() %>%
    split_cols_by_multivar(c("b", "b"), varlabels = c("mean", "range")) %>%
    split_rows_by("a") %>%
    analyze_colvars(list(fun_a, fun_b))

build_table(lyt3, df)
```

gives us

```
             mean          range      
——————————————————————————————————————
a                                     
  readout   0.0401       0.0 - 0.1    
b                                     
  readout    3.51        1.2 - 5.8    
c                                     
  readout   15600    11111.0 - 19992.0
```

So you can see the only things missing on that front are a) a format label specifying signficant digits (I'm not sure when we'll get to this), and more glaringly b) decimal alignment.

Hope that helps.

**Comment by gmbecker** [comment]
When decimal alignment will go in depends on the larger roadmap  (@waddella)

**Comment by legrasv** [comment]
@gmbecker Thank you very much,

@shajoezhu 
I tried it on our PKPT03, it works quite well

```
threesigfmt <- function(x,...) {
  as.character(signif(x, 3))
}

twothreesigfmt <- function(x,...) {
  paste0(as.character(signif(x, 3)), " - ", as.character(signif(x, 3)))
}

fun_a <- function(x, ...) in_rows(readout = mean(x), .formats = list(readout = threesigfmt))
fun_b <- function(x, ...) in_rows(readout = range(x), .formats = list(readout = threesigfmt))


library(scda)
library(rtables)
library(dplyr)
library(tern)

adpp <- synthetic_cdisc_data("latest")$adpp
adpp <- adpp %>% mutate(PKPARAM = factor(paste0(PARAM, " (", AVALU, ")"))) %>%
  filter(PPSPEC == "Plasma", AVISIT == "CYCLE 1 DAY 1")

#lyt creation
lyt <- basic_table() %>%
  split_rows_by(
    var = "ARMCD",
    split_fun = trim_levels_in_group("ARMCD"),
    label_pos = "topleft",
    split_label = "Treatment Arm"
  ) %>%
  split_rows_by(
    var = "TLG_DISPLAY",
    label_pos = "topleft",
    split_label = "PK Parameter"
  ) %>%  
  split_cols_by_multivar(c("AVAL","AVAL"), varlabels = c("mean", "min-max")) %>%
  analyze_colvars(list(fun_a, fun_b))

adpp0 <- adpp %>% filter(PPCAT == "Plasma Drug X") %>% h_pkparam_sort()
build_table(lyt, adpp0)
```
![image](https://user-images.githubusercontent.com/99191912/168766832-0d39541e-9fd6-452b-b418-7913dc6cfede.png)

**Comment by Melkiades** [comment]
> Hi @gmbecker , do you know how to apply 3 sig. figures in `summarize_vars_in_cols`? 

@6iris6 actually the solution was quite simple and the workaround specified by @gmbecker worked perfectly. See the new test here https://github.com/insightsengineering/tern/pull/756, which has the most complete template in it. Still missing alignment (@ayogasekaram) and shifting from summarize to analyze (I am working on it).

**Comment by edelarua** [comment]
Will implement significant figures in `tern` - see https://github.com/insightsengineering/tern/pull/1048. See https://github.com/insightsengineering/formatters/issues/42 for decimal alignment solution.
